src/imperative/imperative.cc

```diff
@@ -273,10 +288,132 @@ void Imperative::RecordOp(
+    s.outputs.emplace_back(ndoutput->deferredcompute_entry_);
+  }
+  std::unordered_map<NDArray *, nnvm::ObjectPtr> ndinput_to_variable;
+  std::unordered_set<const NDArray *> missing_inputs;
+  auto add_symbol_variables = [&inputs, &ndinput_to_variable,
+                               &missing_inputs](const nnvm::ObjectPtr &node) {
+    if (node == nullptr) {
+      // This (nonexistant) "Node" belongs to an array created outside of deferred compute scope.
+      return;
+    }
+
+    // Check if node has any non-deferred compute inputs
+    for (uint32_t i = 0; i < node->inputs.size(); i++) {
+      nnvm::NodeEntry &node_entry = node->inputs[i];
+      if (node_entry.node == nullptr || node_entry.node->is_variable()) {
+        // Node has non-deferred compute input (nullptr). Find the corresponding
+        // NDArray and create a variable for it. If GetDeferredComputeSymbol has
+        // been called before, a variable already exists and only the name needs
+        // to be updated.
+        Imperative::DCInfo &dcinfo = Imperative::DCInfo::Get(node);
+        const NDArray *array = dcinfo.input_handles_.at(i);
+
+        // Make sure this array is part of GetDeferredComputeSymbol inputs
+        auto is_equal = [array](const std::pair<NDArray *, std::string> &input) {
+          return array == std::get<0>(input);
+        };
+
+        // std::vector<std::pair<NDArray *, std::string>>::iterator input_search =
+        auto input_search = std::find_if(inputs.begin(), inputs.end(), is_equal);
+        // Create symbol variable
+        if (input_search != inputs.end()) {
+          NDArray *ndinput;
+          std::string input_name;
+          std::tie(ndinput, input_name) = *input_search;
+
+          nnvm::ObjectPtr input_variable;
+
+          auto variable_search = ndinput_to_variable.find(ndinput);
+          if (variable_search == ndinput_to_variable.end()) {
+            // No variable for this ndarray yet
+            input_variable = nnvm::CreateVariableNode(input_name);
+            ndinput_to_variable.insert({ndinput, input_variable});
+          } else {
+            input_variable = variable_search->second;
+          }
+
+          node_entry.node = input_variable;
+        } else {
+          missing_inputs.insert(array);
+        }
+      }
+    }
+  };
+  nnvm::DFSVisit(s.outputs, add_symbol_variables);
+
+  CHECK_EQ(ndinput_to_variable.size(), inputs.size())
+      << "ValueError: Invalid input to GetDeferredComputeSymbol. "
+      << ndinput_to_variable.size() << " inputs are required, but "
+      << inputs.size() << " were specified.";
+
+  CHECK_EQ(missing_inputs.size(), 0)
+      << "ValueError: Invalid input to GetDeferredComputeSymbol. "
+      << missing_inputs.size() << " required inputs unspecified.";
+
+  // Deep copy of symbol as subsequent calls to this function may change the
+  // name of input variables.
+  s = s.Copy();
+  Symbol *out = new Symbol();
```

Comment:
Related to the above comment, it's better to move this to the caller and simply do `return s.Copy()` here to prevent potential memory leak when an exception is thrown in the following line.